### PR TITLE
feat(foldtext): add option to get virt texts of arbitrary rows

### PR DESCRIPTION
**Pull request by aileot**
The option converts the first argument of virt text handler into a list of virt texts on Lua metatable.
It's useful when the first lines of folds give little information.

For example, we can see the second lines that show plugin names managed by plugin manager at a glance:

1. [packer.nvim](https://github.com/wbthomason/packer.nvim)

```
use {  'kevinhwang91/nvim-ufo',  2
}
```

for

```lua
use {
    'kevinhwang91/nvim-ufo',
    requires = 'kevinhwang91/promise-async'
}
```

2. [dein.vim](https://github.com/Shougo/dein.vim) in toml

```
[[plugins]]  repo = 'kevinhwang91/nvim-ufo'  2
```

for

```toml
[[plugins]]
repo = 'kevinhwang91/nvim-ufo'
depends = 'kevinhwang91/promise-async'
```

A sample handler in which folded lines will be concatenated until each `virtText` becomes longer than `minWidth`:

```lua
--- A list of [`text`, `highlight`] tuples.
---@alias VirtText string[][]

---@param virtText VirtText
---@return number
local virtTextWidth = function(virtText)
  local width = 0
  for _, c in ipairs(virtText) do
    local chunkText = c[1]
    width = width + vim.fn.strdisplaywidth(chunkText)
  end
  return width
end

---@param virtText VirtText
---@return VirtText
local trimIndent = function(virtText)
  local newVirtText = vim.deepcopy(virtText)
  local firstChunkText = newVirtText[1][1]
  newVirtText[1][1] = firstChunkText:gsub("^%s+", "")
  return newVirtText
end

---@param virtText VirtText
---@return VirtText
local compressSpaces = function(virtText)
  local newVirtText = vim.deepcopy(virtText)
  for _, c in ipairs(newVirtText) do
    local chunkText = c[1]
    c[1] = chunkText:gsub("%s+", " ")
  end
  return newVirtText
end

--- Combine two virtual texts.
---@param a VirtText
---@param b VirtText
---@return VirtText
local combineVirtTexts = function(a, b)
  local newVirtText = vim.deepcopy(a)
  for _, chunk in ipairs(b) do
    table.insert(newVirtText, chunk)
  end
  return newVirtText
end

--- Concatenate multiple virtual texts.
---@param a VirtText
---@vararg VirtText
---@return VirtText
local concatVirtTexts = function(a, ...)
  local newVirtText = vim.deepcopy(a)
  for _, b in ipairs({ ... }) do
    newVirtText = combineVirtTexts(newVirtText, b)
  end
  return newVirtText
end

local handler = function(virtTexts, lnum, endLnum, width, truncate)
  local virtText = virtTexts[lnum]
  local minWidth = 25
  local offset = 0
  local separator = { { "  ", "Delimiter" } }
  local maxOffset = endLnum - lnum
  while virtTextWidth(virtText) < minWidth and offset <= maxOffset do
    offset = offset + 1
    virtText = concatVirtTexts(
      virtText,
      separator,
      compressSpaces(trimIndent(virtTexts[lnum + offset]))
  end
  local newVirtText = {}
  local suffix = ('  %d '):format(endLnum - lnum)
  -- The rest would be the same as the example in README.md.
end

require('ufo').setup({
    enable_on_demand_virt_texts = true -- The new option. (default: false)
    -- enable_fold_end_virt_text = true, -- It could be deprecated.
    fold_virt_text_handler = handler
})
```

In addition, I also suggest such functions to deal with `virtText`s as above be provided for users by, for instance, `require('ufo.user-utils')`.

### CODE REVIEW

**Comment by kevinhwang91** [comment]
Not bad, but should pass the metatable into 6th parm for the handler instead of making 1st parm confusing for users.

For instance, if 3rd plugin author needs to integrate ufo with his buffer, he will call `setFoldVirtTextHandler(bufnr, handler)`, but the 1st type of parm depends on ufo's config, it's so confusing.

BTW, I don't know why the name is `enable_on_demand_virt_texts`, could you explain patiently?

File: lua/ufo/decorator.lua [hunk 1]
```diff
@@ -90,13 +90,20 @@ local function onEnd(name, tick)
...
 
                         local handler = self:getVirtTextHandler(bufnr)
-                        local virtText = render.getVirtText(bufnr, text, width, lnum, syntax, nss)
+                        local virtTexts = config.enable_on_demand_virt_texts
+                            and setmetatable({}, {
+                                __index = function(_, _lnum)
+                                    local _text = fb:lines(_lnum)[1]
+                                    return render.getVirtText(bufnr, _text, width, _lnum, syntax, nss)
+                                end,
+                            })
```
Comment (kevinhwang91) [inline lua/ufo/decorator.lua:99]: Should check `_text` is valid.

File: lua/ufo/decorator.lua [hunk 2]
```diff
@@ -90,13 +90,20 @@ local function onEnd(name, tick)
...
+                            or render.getVirtText(bufnr, text, width, lnum, syntax, nss)
                         local endVirtText
                         if self.enableFoldEndVirtText then
                             local endText = fb:lines(endLnum)[1]
                             endVirtText = render.getVirtText(bufnr, endText, width, endLnum, syntax, nss)
                         end
-                        virtText = handler(virtText, lnum, endLnum, width, utils.truncateStrByWidth, {
+                        local virtText = handler(virtTexts, lnum, endLnum, width, utils.truncateStrByWidth, {
```
Comment (kevinhwang91) [inline lua/ufo/decorator.lua:106]: should add the metatable into 6th parm like `end_virt_text`

**Review by kevinhwang91 — CHANGES_REQUESTED** [review]

File: lua/ufo/decorator.lua [hunk 3]
```diff
@@ -90,13 +90,20 @@ local function onEnd(name, tick)
                         local endLnum = utils.foldClosedEnd(0, lnum)
 
                         local handler = self:getVirtTextHandler(bufnr)
-                        local virtText = render.getVirtText(bufnr, text, width, lnum, syntax, nss)
+                        local virtTexts = config.enable_on_demand_virt_texts
+                            and setmetatable({}, {
+                                __index = function(_, _lnum)
+                                    local _text = fb:lines(_lnum)[1]
```
Comment (kevinhwang91) [inline lua/ufo/decorator.lua:96]: use index of a list whose size is `endLnum - lnum` instead of `_lnum`, shouldn't export extra detail.

**Comment by aileot** [comment]
> BTW, I don't know why the name is enable_on_demand_virt_texts, could you explain patiently?

`enable_virt_text_list` the first alternative I think of sounds as if extra calculations have run for every lines in fold to make `virt_text` lists before handler while "on-demand" explicitly demands no unnecessary calculations. The doubt of extra resource consumption with the former name makes me feel uncomfortable beyond reason. Did this answer your question?

Anyway, you can rename it better. You're the author.

File: lua/ufo/decorator.lua (hunk 3 above)
Comment (aileot) [reply lua/ufo/decorator.lua:98]: Resolved in 94b085428ef8041f7e041801b4b9310f551cab2c

File: lua/ufo/decorator.lua (hunk 2 above)
Comment (aileot) [reply lua/ufo/decorator.lua]: Resolved in 61a2703d88f37850147fcbe276def6d44574f74d

File: lua/ufo/decorator.lua (hunk 1 above)
Comment (aileot) [reply lua/ufo/decorator.lua]: You meant a validation for `_lnum`, right?

**Comment by kevinhwang91** [comment]
After considering return a list can't use `ipairs` and `paris` to iter (`__ipairs` and `__pairs` doesn't exist in Lua 5.1), I think exporting a function will be better. Free feel to share your opinion here before merging.

**Comment by aileot** [comment]
I agree with you. A function is more concise than a list in this case.

I'd like to cut off "func" from the function name. It's redundant. Still, it's hard to rename it better.

- `get_folded_line` is to return a string.
- `get_folded_text` is to return a string.
- `get_folded_node` reminds me of tree-sitter node, which is distinct from what we deal with.
- `get_folded_chunks` is used in help file, but it's plural.
- `get_folded_context` is ambiguous.
- `get_folded_content` implies all the lines in a fold, not a line.

Provisionally, I rename it `get_fold_virt_text`, which comes from the option name. Either this or `get_folded_chunks` seems better in them.

What do you think of it?

**Comment by aileot** [comment]
Sorry for the violation.

**Comment by aileot** [comment]
ufo-dein-toml-sample

May I add the screenshot in README.md?

**Comment by kevinhwang91** [comment]
Don't change the struct of `doc/exmaple`, do-end will increase unnecessary indentations. `enable_get_fold_virt_text_func` should be changed to `enable_get_fold_virt_text`.

> Well, may I add the screenshot in README.md?

Sorry, please don't. I want to keep the README.md as clean as possible for users.

**Comment by aileot** [comment]
No problem?

File: doc/example.lua [hunk 4]
```diff
@@ -147,4 +147,128 @@ local function customizeBufFoldText()
...
+end
+
+local function customizeFoldTextWithFunc()
+    require('ufo').setup({
+        enable_get_fold_virt_text = true,
+        fold_virt_text_handler = another_handler,
+    })
+end
+
+-------------------------------------enable_get_fold_virt_text--------------------------------------
```
Comment (kevinhwang91) [inline doc/example.lua:274]: It export so many imp detail, I don't think it is necessary for users. Users should contact/change the virt text themselves if they know what they are doing.  `inspectVirtTextForFoldedLines` is clear enough for users to make them understand what the feature of `enable_get_fold_virt_text` is and how to use that feature.

**Review by kevinhwang91 — COMMENTED** [review]
So many detail for example

File: doc/example.lua (hunk 4 above)
Comment (aileot) [reply doc/example.lua]: Reverted

**Comment by aileot** [comment]
Is that all?

**Comment by kevinhwang91** [comment]
Thanks, please squash and merge the main.

**Review by kevinhwang91 — APPROVED** [review]

**Comment by kevinhwang91** [comment]
Thanks a lot, I have squashed and merged it.

**Comment by wookayin** [comment]
> enable_get_fold_virt_text_func should be changed to enable_get_fold_virt_text.

The option was renamed to `enable_get_fold_virt_text` (from `enable_get_fold_virt_text_func`) and `doc/example.lua` would be need updated accordingly.

**Pull request merged**